pkg/sqlite/load.go

```diff
@@ -1279,121 +1279,141 @@ func (s *sqlLoader) addBundleProperties(tx *sql.Tx, bundle *registry.Bundle) err
+			FROM channel_entry
+			LEFT OUTER JOIN channel_entry AS replaces
+				ON replaces.entry_id = channel_entry.replaces
+			LEFT OUTER JOIN channel_entry AS replaced_by
+				ON channel_entry.entry_id = replaced_by.replaces
+			WHERE channel_entry.operatorbundle_name = ?
+		UNION
+		SELECT channel_entry.operatorbundle_name, replaces.operatorbundle_name, replaced_by.operatorbundle_name, channel_entry.channel_name, channel_entry.package_name
+			FROM channel_entry
+			JOIN replaces_entry
+				ON replaces_entry.replaces = channel_entry.operatorbundle_name
+			LEFT OUTER JOIN channel_entry AS replaces
+				ON channel_entry.replaces = replaces.entry_id
+			LEFT OUTER JOIN channel_entry AS replaced_by
+				ON channel_entry.entry_id = replaced_by.replaces
+		)
+	SELECT
+		replaces_entry.operatorbundle_name,
+		operatorbundle.version,
+		operatorbundle.bundlepath,
+		GROUP_CONCAT(DISTINCT replaces_entry.channel_name),
+		GROUP_CONCAT(DISTINCT replaces_entry.replaces),
+		GROUP_CONCAT(DISTINCT replaces_entry.replaced_by)
+	FROM replaces_entry
+	LEFT OUTER JOIN operatorbundle
+		ON operatorbundle.name = replaces_entry.operatorbundle_name
+	GROUP BY replaces_entry.operatorbundle_name, replaces_entry.package_name
+	ORDER BY replaces_entry.channel_name, replaces_entry.replaces, replaces_entry.replaced_by`
 
-	for len(next) > 0 {
-		// Pop the next bundle off of the queue
-		bundle := next[0]
-		next = next[1:] // Potentially inefficient queue implementation, but this function is only used when deprecate is called
+	getDefaultChannelHead := `
+		SELECT head_operatorbundle_name FROM channel
+		INNER JOIN package ON channel.name = package.default_channel
+		INNER JOIN channel_entry on channel.package_name = channel_entry.package_name
+		WHERE channel_entry.operatorbundle_name = ?
+		LIMIT 1`
 
-		// Check if next is the head of the defaultChannel
-		// If it is, the defaultChannel would be removed -- this is not allowed because we cannot know which channel to promote as the new default
-		var err error
-		if row := tx.QueryRow(isDefaultChannelHead, bundle); row != nil {
-			err = row.Scan(&sql.NullString{})
-		}
-		if err == nil {
-			// A nil error indicates that next is the default channel head
-			return nil, registry.ErrRemovingDefaultChannelDuringDeprecation
-		} else if err != sql.ErrNoRows {
-			return nil, err
-		}
-
-		rows, err := tx.QueryContext(context.TODO(), getReplacesSkips, bundle)
-		if err != nil {
-			return nil, err
-		}
+	row := tx.QueryRow(getDefaultChannelHead, head)
+	if row == nil {
+		return nil, fmt.Errorf("could not find default channel head for %s", head)
+	}
+	var defaultChannelHead sql.NullString
+	err = row.Scan(&defaultChannelHead)
+	if err != nil {
+		return nil, fmt.Errorf("error getting default channel head for %s: %v", head, err)
+	}
+	if !defaultChannelHead.Valid || len(defaultChannelHead.String) == 0 {
```

Comment:
Would that also solve https://bugzilla.redhat.com/show_bug.cgi?id=2001940?

Reply:
Solved by https://github.com/operator-framework/operator-registry/pull/780/files